### ISSUE TITLE
Confusion computing nested virtual properties

### ISSUE THREAD

**Issue by AllenSH12** [Automattic/mongoose#5010]
**Do you want to request a *feature* or report a *bug*?**
Bug

**What is the current behavior?**
When virtual attributes are nested they appear to collide with attributes defined on other schemas.

**If the current behavior is a bug, please provide the steps to reproduce.**

Test script:
```
#!/usr/bin/env node

var mongoose = require('mongoose');
var Schema = mongoose.Schema;


var fooSchema = new Schema({
  name: String,
  some_object: Object
});
fooSchema.virtual('some_object.bar').get(function() {
  console.log('VIRTUAL ONE');

  return this.name;
});
fooSchema.set('toObject', { virtuals: true });


var barSchema = new Schema({
  name2: String,
  some_object: Object
});
barSchema.virtual('some_object.bar').get(function() {
  console.log('VIRTUAL TWO');

  return this.name2;
});
barSchema.set('toObject', { virtuals: true });


var Foo = mongoose.model('Foo', fooSchema);
var Bar = mongoose.model('Bar', barSchema);

var test = Foo({name: 'orlando'});
var test2 = Bar({name2: 'allen'});

console.log(test.toObject());
console.log(test2.toObject());
```

And the output in the console...
```
VIRTUAL TWO
VIRTUAL ONE
{ name: 'orlando',
  _id: 58ae3d6f23205b29cec088d1,
  some_object: { bar: 'orlando' },
  id: '58ae3d6f23205b29cec088d1' }
VIRTUAL TWO
VIRTUAL ONE
{ name2: 'allen',
  _id: 58ae3d6f23205b29cec088d2,
  some_object: { bar: undefined },
  id: '58ae3d6f23205b29cec088d2' }
```

**What is the expected behavior?**
I would not expect virtual property definitions to have any impact across distinct instances of schemas. The expected output would be as if the field was defined using a non-nested path, i.e. passing the string "foobar" to `.virtual` method as opposed to "some_object.bar"

```
VIRTUAL ONE
{ name: 'orlando',
  _id: 58ae3d6f23205b29cec088d1,
  some_object: { bar: 'orlando' },
  id: '58ae3d6f23205b29cec088d1' }
VIRTUAL TWO
{ name2: 'allen',
  _id: 58ae3d6f23205b29cec088d2,
  some_object: { bar: 'allen' },
  id: '58ae3d6f23205b29cec088d2' }
```

**Please mention your node.js, mongoose and MongoDB version.**
```
node@v6.5.0
mongoose@4.8.4
mongo@v3.0.3 (also observed w/ mongo@v3.2.8)
```

Thank you for your time.

**Comment by AllenSH12** [comment]
Solved by a co-worker, for posterity:

tl;dr, dont use the shorthand constructors

they reference the same types if found at the same path and so are shared across schemas. Mutating one affects all the others (edited)

however, using the longer version you're always safe, like:
```
const schema = {
  some_object: {
    type: Object
  }
};
```
Because the `type` at the path string will always be an object literal which has it's own place in memory and is not shared across schemas. Whereas:
```
const schema = {
  some_object: Object
}
```
has the Object Constructor function saved as it's value at `some_object`
